util/testutil/integration/dockerd.go

```diff
@@ -0,0 +1,125 @@
+package integration
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"time"
+
+	"github.com/docker/docker/testutil/daemon"
+)
+
+const dockerdBinary = "dockerd"
+
+type logTAdapter struct {
+	Name string
+	Logs map[string]*bytes.Buffer
+}
+
+func (l logTAdapter) Logf(format string, v ...interface{}) {
+	if buf, ok := l.Logs[l.Name]; !ok || buf == nil {
+		l.Logs[l.Name] = &bytes.Buffer{}
+	}
+	fmt.Fprintf(l.Logs[l.Name], format, v...)
+}
+
+// InitDockerdWorker registers a dockerd worker with the global registry.
+func InitDockerdWorker() {
+	Register(&dockerd{})
+}
+
+type dockerd struct{}
+
+func (c dockerd) Name() string {
+	return dockerdBinary
+}
+
+func (c dockerd) New(cfg *BackendConfig) (b Backend, cl func() error, err error) {
+	if err := requireRoot(); err != nil {
+		return nil, nil, err
+	}
+
+	deferF := &multiCloser{}
+	cl = deferF.F()
+
+	defer func() {
+		if err != nil {
+			deferF.F()()
+			cl = nil
+		}
+	}()
+
+	workDir, err := ioutil.TempDir("", "integration")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cmd, err := daemon.NewDaemon(
+		workDir,
+		daemon.WithTestLogger(logTAdapter{
+			Name: "creatingDaemon",
+			Logs: cfg.Logs,
+		}),
+		daemon.WithContainerdSocket(""),
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("new daemon error: %q, %s", err, formatLogs(cfg.Logs))
+	}
+
+	err = cmd.StartWithError()
+	if err != nil {
+		return nil, nil, err
+	}
+	deferF.append(cmd.StopWithError)
+
+	logs := map[string]*bytes.Buffer{}
+	if err := waitUnix(cmd.Sock(), 5*time.Second); err != nil {
+		return nil, nil, fmt.Errorf("dockerd did not start up: %q, %s", err, formatLogs(logs))
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	deferF.append(func() error { cancel(); return nil })
+
+	dockerAPI, err := cmd.NewClient()
+	if err != nil {
+		return nil, nil, err
+	}
+	deferF.append(dockerAPI.Close)
+
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return nil, nil, err
+	}
+	deferF.append(listener.Close)
+
+	go func() {
+		for {
+			tmpConn, err := listener.Accept()
+			if err != nil {
+				return
+			}
+			conn, err := dockerAPI.DialHijack(ctx, "/grpc", "h2c", nil)
+			if err != nil {
+				return
+			}
+
+			go func() {
```

Comment:
try to switch these to  https://godoc.org/golang.org/x/sync/errgroup (or https://golang.org/pkg/sync/#WaitGroup)

Reply:
actually, not sure if `errgroup` is needed but logging on errors would be nice

Reply:
Logging would be good, but I don't see a good way to do that right now. Existing logs in this package just get returned and formatted somewhere else. Even if we add some locking around the log aggregation, they have already been returned by the time this goroutine completes, so I'm not sure there's a good place to do it unless you'd like a larger refactor of how logging works? I'd love to get rid of the aggregation and find a way to stream logs somewhere instead.

Reply:
Returned the first error in the `deferF` which I had forgotten about, this should end up causing the test to fail and logging.